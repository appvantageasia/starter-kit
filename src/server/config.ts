import chalk from 'chalk';
import { Region } from 'minio';
import { SMTPSettings } from './emails';
import { getString, getBoolean, getInteger, getNumber } from './env';

const prefix = 'APP';

const getPrefix = key => `${prefix}_${key}`;

export type RuntimeConfig = {
    version: string;

    port: number;

    publicPath: string;

    db: {
        uri: string;
        name: string;
        pool: number;
    };

    redis: {
        uri: string;
    };

    session: {
        secret: string;
        lifetime: string;
    };

    i18n: {
        locales: string[];
        defaultLocale: string;
    };

    smtp: {
        transporter: SMTPSettings;
        sender: string;
    };

    storage: {
        provider: {
            endPoint: string;
            accessKey: string;
            secretKey: string;
            useSSL?: boolean;
            port?: number;
            region: Region;
        };
        bucket: string;
    };

    sentry: {
        dsn?: string;
        release?: string;
        environment?: string;
        tracing: boolean;
        tracesSampleRate: number;
    };

    html2pdf: {
        endpoint: string;
    };
};

const getSmtpSettings = (): RuntimeConfig['smtp']['transporter'] => {
    const base = {
        host: getString(getPrefix('SMTP_HOST'), 'localhost'),
        port: getInteger(getPrefix('SMTP_PORT'), 465),
        secure: getBoolean(getPrefix('SMTP_SECURE'), false),
    };

    const user = getString(getPrefix('SMTP_USER'));

    if (!user) {
        return base;
    }

    return {
        ...base,
        auth: {
            user,
            pass: getString(getPrefix('SMTP_PASSWORD')),
        },
    };
};

const version = getString('VERSION', '0.0.0-development');

const config: RuntimeConfig = {
    version,

    port: getNumber(getPrefix('PORT'), 3000),
    publicPath: getString(getPrefix('PUBLIC_PATH'), '/public/'),

    // internationalization
    i18n: {
        locales: ['en'],
        defaultLocale: 'en',
    },

    // server runtime
    db: {
        uri: getString(getPrefix('DB_URI'), 'mongodb+srv://localhost:27017'),
        name: getString(getPrefix('DB_NAME'), 'app'),
        pool: getInteger(getPrefix('DB_POOL'), 10),
    },

    redis: {
        uri: getString(getPrefix('REDIS_URI'), 'redis://127.0.0.1:6379'),
    },

    session: {
        secret: getString(getPrefix('SESSION_SECRET')),
        lifetime: getString(getPrefix('SESSION_LIFETIME'), '1h'),
    },

    // default/system SMTP settings
    smtp: {
        transporter: getSmtpSettings(),
        sender: getString(getPrefix('SMTP_FROM'), 'noreply@appvantage.co'),
    },

    storage: {
        provider: {
            endPoint: getString(getPrefix('STORAGE_ENDPOINT')),
            accessKey: getString(getPrefix('STORAGE_ACCESS_KEY')),
            secretKey: getString(getPrefix('STORAGE_SECRET_KEY')),
            useSSL: getBoolean(getPrefix('STORAGE_SSL'), true),
            port: getInteger(getPrefix('STORAGE_PORT')),
            region: getString(getPrefix('STORAGE_REGION'), 'ap-southeast-1'),
        },
        bucket: getString(getPrefix('STORAGE_BUCKET'), 'app'),
    },

    sentry: {
        dsn: getString(getPrefix('SENTRY_DSN')),
        release: getString(getPrefix('SENTRY_RELEASE')),
        environment: getString(getPrefix('SENTRY_ENVIRONMENT')),
        tracing: getBoolean(getPrefix('SENTRY_TRACING'), true),
        tracesSampleRate: getNumber(getPrefix('SENTRY_TRACES_SAMPLE_RATE'), 1.0),
    },

    html2pdf: {
        endpoint: getString(getPrefix('HTML2PDF_ENDPOINT')),
    },
};

const exitOnError = (error: string): void => {
    console.error(error);
    process.exit(1);
};

export const runValidityChecks = () => {
    if (!config.session.secret) {
        exitOnError(chalk.red('APP_SESSION_SECRET is missing in environment variables'));
    }
};

export default config;
