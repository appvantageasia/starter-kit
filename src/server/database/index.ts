export * from './documents';
export * from './collections';
export * from './instance';
export * from './migrate';
export * from './helpers';
