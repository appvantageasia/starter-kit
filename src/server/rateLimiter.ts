import { RateLimiterRedis } from 'rate-limiter-flexible';
import getRedisInstance from './redis';

// rate limiter for any web request for other than assets
export const expressRateLimiter = new RateLimiterRedis({
    storeClient: getRedisInstance(),
    points: 10, // 10 requests per second max
    duration: 1, // reset every seconds
    keyPrefix: 'expressRlfx',
});

// rate limiter for authentication try outs
// over 5 tries per minute for a username or ip
export const authenticationRateLimiter = new RateLimiterRedis({
    storeClient: getRedisInstance(),
    points: 10, // 5 tries per minute max, tries on username consume 2 points
    duration: 60, // reset every minutes
    keyPrefix: 'authRlfx',
    blockDuration: 3 * 60, // block for 3 minutes
});
