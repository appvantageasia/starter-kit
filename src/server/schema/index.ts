import { makeExecutableSchema } from '@graphql-tools/schema';
import resolvers from './resolvers';
import typeDefs from './typeDefs.graphql';

const schema = makeExecutableSchema({ typeDefs: [typeDefs], resolvers });

export default schema;
