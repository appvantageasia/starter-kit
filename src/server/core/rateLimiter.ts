import { RateLimiterRedis } from 'rate-limiter-flexible';
import config from './config';
import getRedisInstance from './redis';

// rate limiter for any web request for other than assets
export const expressRateLimiter = new RateLimiterRedis({
    storeClient: getRedisInstance(),
    points: config.limiter.api, // n requests per second max
    duration: 1, // reset every seconds
    keyPrefix: 'expressRlfx',
});

// rate limiter for authentication try outs
// over 5 tries per minute for a username or ip
export const authenticationRateLimiter = new RateLimiterRedis({
    storeClient: getRedisInstance(),
    points: 10, // 5 tries per minute max, tries on username consume 2 points
    duration: 60, // reset every minutes
    keyPrefix: 'rflx:auth',
    blockDuration: 3 * 60, // block for 3 minutes
});

// rate limiter for link
export const linkRetrievalRateLimiter = new RateLimiterRedis({
    storeClient: getRedisInstance(),
    points: 1, // only 1 try per second
    duration: 1,
    keyPrefix: 'rflx:linkRetrieval',
});

// rate limiter on reset password
export const resetPasswordRateLimiter = new RateLimiterRedis({
    storeClient: getRedisInstance(),
    points: 1, // only 1 try per second
    duration: 1,
    keyPrefix: 'rflx:resetPassword',
});

// rate limiter for TOTP
export const totpRateLimiter = new RateLimiterRedis({
    storeClient: getRedisInstance(),
    points: 1, // only 1 try per second
    duration: 1,
    keyPrefix: 'totpRflex',
});

// rate limiter to publically fetch webauthn keys
export const webAuthnKeyFetchRateLimiter = new RateLimiterRedis({
    storeClient: getRedisInstance(),
    points: 1, // only 1 try per second
    duration: 1,
    keyPrefix: 'rflx:webAuthnKeyFetch',
});
