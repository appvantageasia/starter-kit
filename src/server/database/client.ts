import { MongoClient, MongoClientOptions } from 'mongodb';
import config from '../core/config';
import { getEncryptionKeyId } from './helpers';
import getKMS from './kms';
import buildSchemaMap from './schemaMap';

export const getRegularOptions = (): MongoClientOptions => ({
    maxPoolSize: config.db.pool,
});

export const getRegularClient = () => MongoClient.connect(config.db.uri, getRegularOptions());

export const getEncryptedClient = async (kms: Exclude<ReturnType<typeof getKMS>, null>) => {
    const keyId = await getEncryptionKeyId();

    if (!keyId) {
        throw new Error('No encryption key ID have been provided for the schema');
    }

    return MongoClient.connect(config.db.uri, {
        ...getRegularOptions(),

        // enable command monitoring for the client
        monitorCommands: true,

        autoEncryption: {
            keyVaultNamespace: kms.keyVaultNamespace,
            kmsProviders: kms.kmsProviders,
            schemaMap: buildSchemaMap(keyId),

            // configure cryptd client
            extraOptions: {
                mongocryptdURI: config.db.cryptd.uri,
                mongocryptdBypassSpawn: config.db.cryptd.mongocryptdBypassSpawn,
                mongocryptdSpawnArgs: config.db.cryptd.mongocryptdSpawnArgs,
            },
        },
    });
};
