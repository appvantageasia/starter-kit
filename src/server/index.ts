import { program } from 'commander';
import {
    composeCommand,
    executeDataMigration,
    startServerCommand,
    startWorkerCommand,
    startBullBoardCommand,
    executeSetupMasterKey,
} from './core/commands';
import config, { runValidityChecks } from './core/config';
import createWebServer from './core/createWebServer';
import { HealthStatusManager, createHealthServer } from './core/health';
import { startPrometheusServer } from './core/prometheus';
import { listPendingMigrations } from './database';
import { setup as startWorker } from './queues';

runValidityChecks();

program.version(config.version);

program
    .command('migrate')
    .description('Execute data migration')
    .action(() => executeDataMigration(true));

program
    .command('worker')
    .description('Start worker')
    .action(() => {
        const manager = new HealthStatusManager();
        createHealthServer(manager);
        composeCommand(startWorkerCommand(manager));
    });

program
    .command('serve')
    .description('Start web server')
    .action(async () => {
        const manager = new HealthStatusManager();
        createHealthServer(manager);
        composeCommand(await startServerCommand(manager), await startPrometheusServer());
    });

program
    .command('bullBoard')
    .description('Start bull board')
    .action(async () => {
        const manager = new HealthStatusManager();
        createHealthServer(manager);
        composeCommand(startBullBoardCommand(manager));
    });

program.command('setupMasterKey').description('Setup master key for MongoDB CSFLE').action(executeSetupMasterKey);

if (process.isCLI) {
    program.parse();
}

const bundle = { createWebServer, startWorker, executeDataMigration, listPendingMigrations };

export default bundle;

export type Bundle = typeof bundle;
