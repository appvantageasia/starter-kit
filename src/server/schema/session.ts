import { IncomingMessage } from 'http';
import { AuthenticationError } from 'apollo-server';
import cookie from 'cookie';
import jwt from 'jsonwebtoken';
import { getOr } from 'lodash/fp';
import { ObjectId } from 'mongodb';
import config from '../core/config';
import { generateToken, consumeToken, csrfGenerator } from '../core/tokens';
import { getDatabaseContext } from '../database';
import getIp from '../utils/getIp';

export type SessionData = {
    userId: ObjectId;
    sessionId: ObjectId;
};

export const getSessionToken = async (data: SessionData) => {
    const csrf = csrfGenerator();
    const token = generateToken('session', data, config.session.lifetime, csrf);

    // get iat and exp from token instead
    const { iat, exp } = jwt.decode(token, { json: true });

    return {
        token: generateToken('session', data, config.session.lifetime, csrf),
        csrf,
        exp: new Date(exp * 1000),
        iat: new Date(iat * 1000),
    };
};

export const readSessionToken = (token: string, csrf: string) => consumeToken<SessionData>('session', token, csrf);

export const getCSRFFromHeaders = (req: IncomingMessage): string => {
    const requestCookie = req.headers?.cookie;

    return requestCookie ? getOr('', 'CSRF', cookie.parse(requestCookie)) : '';
};

const getAuthorizationFromHeader = (req: IncomingMessage) => {
    const header = req.headers?.authorization as string;

    if (!header) {
        return null;
    }

    const [type, token] = header.split(' ');

    if (type !== 'Bearer') {
        throw new AuthenticationError('Unsupported authentication method');
    }

    if (!token) {
        throw new AuthenticationError('Missing token');
    }

    return token;
};

export const getSessionDataFromRequest = async (
    req: IncomingMessage,
    authorization?: string
): Promise<SessionData | null> => {
    const token = authorization || getAuthorizationFromHeader(req);

    if (!token) {
        return null;
    }

    const data = await (async () => {
        try {
            const csrf = getCSRFFromHeaders(req);

            return await readSessionToken(token, csrf);
        } catch (error) {
            throw new AuthenticationError('Invalid authentication token');
        }
    })();

    if (!data) {
        return null;
    }

    const { collections } = await getDatabaseContext();
    const result = await collections.userSessions.updateOne(
        { _id: data.sessionId },
        {
            $set: {
                // update user agent to keep the latest
                userAgent: req.headers['user-agent'],
                ip: getIp(req),
                // then update last activity date
                lastActivityAt: new Date(),
            },
        }
    );

    if (!result.matchedCount) {
        // session was invalidated
        throw new AuthenticationError('Session invalidated');
    }

    return data;
};
