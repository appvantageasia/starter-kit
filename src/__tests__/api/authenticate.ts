import { ApolloError, gql } from '@apollo/client';
import Dayjs from 'dayjs';
import { authenticator } from 'otplib';
import { authenticationRateLimiter } from '../../server/core/rateLimiter';
import { getDatabaseContext, OTPKind } from '../../server/database';
import { readSessionToken } from '../../server/schema/session';
import {
    cleanDatabase,
    composeHandlers,
    getApolloClient,
    loadFixtures,
    setupDatabase,
    setupWebService,
} from '../helpers';
import fixtures from './authenticate.fixture.json';

const authenticateMutation = gql`
    mutation test($username: String!, $password: String!) {
        response: authenticate(username: $username, password: $password) {
            __typename

            ... on AuthenticationSuccessful {
                user {
                    id
                }
                token
            }

            ... on AuthenticationRequiresTOTP {
                token
            }

            ... on AuthenticationRequiresNewPassword {
                token
            }
        }
    }
`;

const totpMutation = gql`
    mutation test($token: String!, $password: String!) {
        response: authenticateWithTOTP(token: $token, password: $password) {
            __typename

            ... on AuthenticationSuccessful {
                user {
                    id
                }
                token
            }

            ... on AuthenticationRequiresNewPassword {
                token
            }
        }
    }
`;

const passwordChangeMutation = gql`
    mutation test($token: String!, $password: String!) {
        response: changePasswordFromAuthentication(token: $token, password: $password) {
            __typename

            ... on AuthenticationSuccessful {
                user {
                    id
                }
                token
            }
        }
    }
`;

const resetRateLimiter = async () => {
    await authenticationRateLimiter.delete('::ffff:127.0.0.1');
    await authenticationRateLimiter.delete('x');
    await authenticationRateLimiter.delete('unknown');
};

const webService = setupWebService();

beforeEach(composeHandlers(setupDatabase, loadFixtures(fixtures), webService.initialize, resetRateLimiter));

afterEach(composeHandlers(cleanDatabase, webService.cleanUp));

test('Authentication rejects invalid credentials because the user does not exist', async () => {
    const { client } = getApolloClient(webService.url);
    const variables = { username: 'unknown', password: 'something' };
    const promise = client.mutate({ mutation: authenticateMutation, variables });
    await expect(promise).rejects.toBeInstanceOf(ApolloError);
    const error: ApolloError = await promise.catch(error => error);
    expect(error.graphQLErrors).toMatchSnapshot();
});

test('Authentication rejects invalid credentials because the password does not match', async () => {
    const { client } = getApolloClient(webService.url);
    const variables = { username: 'x', password: 'something' };
    const promise = client.mutate({ mutation: authenticateMutation, variables });
    await expect(promise).rejects.toBeInstanceOf(ApolloError);
    const error: ApolloError = await promise.catch(error => error);
    expect(error.graphQLErrors).toMatchSnapshot();
});

test('Authentication returns the user and token on valid credentials', async () => {
    const { collections } = await getDatabaseContext();
    await collections.users.updateOne({ username: 'x' }, { $set: { passwordFrom: new Date() } });
    const { client, getCSRF } = getApolloClient(webService.url);
    const variables = { username: 'x', password: 'y' };
    const { data } = await client.mutate({ mutation: authenticateMutation, variables });
    expect(data.response.__typename).toEqual('AuthenticationSuccessful');
    expect(data.response.user).toMatchSnapshot();
    const sessionData = await readSessionToken(data.response.token, getCSRF());
    expect(sessionData.userId).toMatchSnapshot();
});

test('Authentication returns a request to renew password after expiry', async () => {
    const { collections } = await getDatabaseContext();
    const date = Dayjs().subtract(91, 'day').toDate();
    await collections.users.updateOne({ username: 'x' }, { $set: { passwordFrom: date } });
    const { client } = getApolloClient(webService.url);
    const variables = { username: 'x', password: 'y' };
    const { data } = await client.mutate({ mutation: authenticateMutation, variables });
    expect(data.response.__typename).toEqual('AuthenticationRequiresNewPassword');
    expect(data.response.token).not.toBeNull();
});

test('Authentication returns a request for TOTP when enabled before password expiry check', async () => {
    const { collections } = await getDatabaseContext();
    const date = Dayjs().subtract(91, 'day').toDate();
    const totpSecret = authenticator.generateSecret();
    await collections.users.updateOne(
        { username: 'x' },
        { $set: { passwordFrom: date, otpSetup: { kind: OTPKind.TOTP, secret: totpSecret } } }
    );
    const { client } = getApolloClient(webService.url);
    const variables = { username: 'x', password: 'y' };
    const { data } = await client.mutate({ mutation: authenticateMutation, variables });
    expect(data.response.__typename).toEqual('AuthenticationRequiresTOTP');
    expect(data.response.token).not.toBeNull();
});

test('Authentication returns a request for TOTP when enabled and when password is still valid', async () => {
    const { collections } = await getDatabaseContext();
    const totpSecret = authenticator.generateSecret();
    await collections.users.updateOne(
        { username: 'x' },
        { $set: { passwordFrom: new Date(), otpSetup: { kind: OTPKind.TOTP, secret: totpSecret } } }
    );
    const { client } = getApolloClient(webService.url);
    const variables = { username: 'x', password: 'y' };
    const { data } = await client.mutate({ mutation: authenticateMutation, variables });
    expect(data.response.__typename).toEqual('AuthenticationRequiresTOTP');
    expect(data.response.token).not.toBeNull();
});

test('Authentication is successful with TOTP', async () => {
    const { collections } = await getDatabaseContext();
    const totpSecret = authenticator.generateSecret();
    const passwordFrom = new Date();
    await collections.users.updateOne(
        { username: 'x' },
        { $set: { passwordFrom, otpSetup: { kind: OTPKind.TOTP, secret: totpSecret } } }
    );
    const { client, getCSRF } = getApolloClient(webService.url);
    const { data: authenticateData } = await client.mutate({
        mutation: authenticateMutation,
        variables: { username: 'x', password: 'y' },
    });
    expect(authenticateData.response.__typename).toEqual('AuthenticationRequiresTOTP');
    expect(authenticateData.response.token).not.toBeNull();
    const { data: totpData } = await client.mutate({
        mutation: totpMutation,
        variables: { token: authenticateData.response.token, password: authenticator.generate(totpSecret) },
    });
    expect(totpData.response.__typename).toEqual('AuthenticationSuccessful');
    expect(totpData.response.user).toMatchSnapshot();
    const sessionData = await readSessionToken(totpData.response.token, getCSRF());
    expect(sessionData.userId).toMatchSnapshot();
});

test('Authentication is successful with TOTP and password renewal', async () => {
    const { collections } = await getDatabaseContext();
    const totpSecret = authenticator.generateSecret();
    const passwordFrom = Dayjs().subtract(91, 'day').toDate();
    await collections.users.updateOne(
        { username: 'x' },
        { $set: { passwordFrom, otpSetup: { kind: OTPKind.TOTP, secret: totpSecret } } }
    );
    const { client, getCSRF } = getApolloClient(webService.url);
    const { data: authenticateData } = await client.mutate({
        mutation: authenticateMutation,
        variables: { username: 'x', password: 'y' },
    });
    expect(authenticateData.response.__typename).toEqual('AuthenticationRequiresTOTP');
    expect(authenticateData.response.token).not.toBeNull();
    const { data: totpData } = await client.mutate({
        mutation: totpMutation,
        variables: { token: authenticateData.response.token, password: authenticator.generate(totpSecret) },
    });
    expect(totpData.response.__typename).toEqual('AuthenticationRequiresNewPassword');
    expect(totpData.response.token).not.toBeNull();
    const { data: passwordData } = await client.mutate({
        mutation: passwordChangeMutation,
        variables: { token: totpData.response.token, password: 'S%1209!jagqo712oPI' },
    });
    expect(passwordData.response.__typename).toEqual('AuthenticationSuccessful');
    expect(passwordData.response.user).toMatchSnapshot();
    const sessionData = await readSessionToken(passwordData.response.token, getCSRF());
    expect(sessionData.userId).toMatchSnapshot();
});
