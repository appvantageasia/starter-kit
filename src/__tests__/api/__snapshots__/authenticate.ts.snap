// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication is successful with TOTP 1`] = `
{
  "__typename": "User",
  "id": "601f7ae763d6cfc2554f6b67",
}
`;

exports[`Authentication is successful with TOTP 2`] = `"601f7ae763d6cfc2554f6b67"`;

exports[`Authentication is successful with TOTP and password renewal 1`] = `
{
  "__typename": "User",
  "id": "601f7ae763d6cfc2554f6b67",
}
`;

exports[`Authentication is successful with TOTP and password renewal 2`] = `"601f7ae763d6cfc2554f6b67"`;

exports[`Authentication rejects invalid credentials because the password does not match 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "password": "Invalid credentials",
    },
    "locations": [
      {
        "column": 3,
        "line": 2,
      },
    ],
    "message": "bad request",
    "path": [
      "response",
    ],
  },
]
`;

exports[`Authentication rejects invalid credentials because the user does not exist 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
      "password": "Invalid credentials",
    },
    "locations": [
      {
        "column": 3,
        "line": 2,
      },
    ],
    "message": "bad request",
    "path": [
      "response",
    ],
  },
]
`;

exports[`Authentication returns the user and token on valid credentials 1`] = `
{
  "__typename": "User",
  "id": "601f7ae763d6cfc2554f6b67",
}
`;

exports[`Authentication returns the user and token on valid credentials 2`] = `"601f7ae763d6cfc2554f6b67"`;
